Editor: deploy the new insert menu along with a Guided Tour

Copons opened the pull request:
Add the relevant feature to the production configuration.
Add a brief Guided Tour for this new feature.

![screen shot 2016-11-29 at 18 37 32](https://cloud.githubusercontent.com/assets/2070010/20723627/2c2563c6-b663-11e6-84e6-d7c60ddf2fe0.png)

## Concerns and Doubts

- `<Tour>` currently only allows one `path` (which indicate where to show the tour itself). In this case, though, the tour should appear on both `/post` and `/page`.
Would it be better to change the `path` prop to accept a list of paths, or I'd be better off creating two separate GT components?

- Hack 1: GTs currently require at least two `<Step>`s to work, or [this](https://github.com/Automattic/wp-calypso/blob/master/client/layout/guided-tours/config-elements.js#L69-L70) fails.
As this is an edge case (a tour _should_ need more than one step indeed!), I simply added an empty step and called it a day. It's not elegant, but it's a light enough solution that any impact would be negligible.

- Hack 2: the insert menu button is part of TinyMCE which usually renders after the GT component. So, the GT appears in the wrong position, and at some re-render moves into the right one.
I hated this so much that I added an invisible anchor in the editor header to tether the GT step - and this way I could also position the step precisely wherever I wanted.
It looks much better but it also needs a new element with no other purpose except positioning a temporary component. Any thoughts on this?
FYI this is how it's aligned when positioned relatively to the actual insert menu button:
![screen shot 2016-11-29 at 19 13 30](https://cloud.githubusercontent.com/assets/2070010/20724903/fa8b2b5c-b667-11e6-9aa8-df2d09c5c2fb.png)


- I'm not sure what to do with the `when` prop. Is there a function to make it work only once or to determine the first time a user opens a path?

folletto commented:
> (a tour should need more than one step indeed!)

Errrr not really? A tour of one step is totally legit. :) 

> the insert menu button is part of TinyMCE which usually renders after the GT component.

Can you maybe just anchor it top-left to the whole TinyMCE?

---

Awesome work! This is a great use of the tour!

Copons commented:
@folletto 

> Errrr not really? A tour of one step is totally legit. :)

One-step tour = popover, though. 😆 

_No, I won't even try tethering a popover to a TinyMCE button..._

> Can you maybe just anchor it top-left to the whole TinyMCE?

Top-left from the left side? I guess I could do it, but I'm trying to point to the arrow part of the button as opposed to the + icon, as that's actually the core of the change.

folletto commented:
> One-step tour = popover, though. 

Yes, but GuidedTours include trigger logic, they are linkable, and provide a lot of other support tools. The big part of tours isn't the visual, it's the hidden logic. :) 


> Top-left from the left side? I guess I could do it, but I'm trying to point to the arrow part of the button as opposed to the + icon, as that's actually the core of the change.

I'm not entirely sure it's super clear either way, but yes, I understand.

mcsf commented:
ohai, nice to see you working on this!

> `<Tour>` currently only allows one `path`

Right. So what we've done when building mock tours is use the wildcard `path="/"` and instead rely on `when` to make sure we're in the right section. That is hacky, though, obviously, and yours seems like a good use case for a more flexible `path` prop.

`path` is used to build the [`relevantFeatures` dict](https://github.com/Automattic/wp-calypso/blob/b5137d232929c09e398794227c3dbee2299b7f76/client/state/ui/guided-tours/selectors/index.js#L41). Off the top of my mind, I would replace the `map` call with a `flatMap` and have each iteratee return a list:

```js
( tourMeta, key ) => castToArray( tourMeta.path ).map( path => ( {
  tour: key,
  when: tourMeta.when
  path,
} ) )
```

So that, in the end, `relevantFeatures` has the exact same shape, but now we support `String | Array<String>` for `path`.

> GTs currently require at least two `<Step>`s to work

That's… unexpected. That snippet you linked to isn't actually expecting more than one step, though: it's attempting to grab the very first step. Could the bug arise from the way [React differently handles](https://facebook.github.io/react/docs/react-api.html#react.children) one-element `children` vs. multi-element `children`?

> I hated this so much that I added an invisible anchor in the editor header to tether the GT step […] Any thoughts on this?

Honestly, that sort of issue is probably always going to arise when working with GT, just due to how it's built. More than anything, I'm glad to have someone on the outside trying to use it and come up with solutions. In short: I have no qualms there, experiment at will. :)

> I'm not sure what to do with the `when` prop. Is there a function to make it work only once or to determine the first time a user opens a path?

`when` is definitely one that will need proper documentation. The key thing is that `when` can be different things, depending on what element it is a prop of: `Tour`, `Step`, or `Continue`. If you're referring to `Tour`'s, then `when` is used as the [final starting condition for a tour](https://github.com/Automattic/wp-calypso/blob/b5137d232929c09e398794227c3dbee2299b7f76/client/state/ui/guided-tours/selectors/index.js#L141) if the user has recently navigated to that tour's `path` — that is, `when( state )` is called once to see if we can start a tour, but we don't check again during the tour. This is in contrast with other uses of `when` (in `Step` and `Continue`).

^ the above sounds confusing, so feel free to ask for a better explanation!

Copons commented:
@mcsf 
> Could the bug arise from the way React differently handles one-element children vs. multi-element children?

Yup, spot on.
I didn't notice single child elements were handled as objects.
I'm gonna open a PR protecting the Tour render from that possibility, and hopefully we'll get it merged before this other one.

mcsf commented:
> Hack 2: the insert menu button is part of TinyMCE which usually renders after the GT component. So, the GT appears in the wrong position, and at some re-render moves into the right one.

@Copons I totally forgot to give you this tip: a Step's `target` doesn't have to be a `tip-target` slug, it can be a `querySelector`-type selector *if* the first character is a `.`:

```js
<Step target=".my-component .my-item" />
```

Why is this relevant? Because "ready" states can be encoded in DOM classes. This is [what we did](https://github.com/Automattic/wp-calypso/pull/9610/files#diff-078d1244a4582deb4b57c91578b00380R73) in the concept `main` tour:

`target=".web-preview.is-visible [data-tip-target='web-preview__close']"`

Without the `.is-visible` part, the step will be rendered incorrectly.

mcsf commented:
> `<Tour>` currently only allows one `path`

#9761

Copons commented:
@mcsf I'll have to check how TinyMCE's CSS changes before and after rendering, but meanwhile I've noticed a couple more positioning problems.

- If the editor switches to code view, the guided tour stays there even though its (real) target disappeared.
- If the post has a featured image, there's the possibility that the GT renders at the right position, but then the editor gets pushed down by the image and the GT stays still (until a scroll, at least).

I'm gonna investigate how all those parts move during loading.

Copons commented:
Regarding the positioning issues: I've removed the invisible anchor in the post header, which wasn't enough in case the post had a featured image.

I've replaced it with a new `RepositioningStep` class that inherits everything from `Step` but forces a `onScrollOrResize()` every couple of seconds.

I'm aware this is quite bad, but my reasoning is that the Step covers the editor toolbar and part of the editor body, so I believe it will be dismissed quickly, clearing the interval.

lsinger commented:
To delay rendering, I just realized you can do the following:

```
<RepositioningStep
    ...
    style={ { animationDelay: '10s' } }
    ...
    >
```

When I opened the tour on a blog post that had a featured image and several other images (so it was slow to load and jumped around erratically for a few seconds) using that longer delay made a huge difference. With the 10 second delay there was no jumping around whatsoever. It looked really nice. 

Of course for other users it will be best to have 11 seconds, or 12 seconds, and so on. This is just to say that I figured out a way to delay the initial rendering of the tour by a constant number of seconds.

lsinger commented on client/layout/guided-tours/editor-insert-menu-tour.js:
```diff
@@ -0,0 +1,76 @@
+		>
+			<p>
+				{ translate(
+					'{{strong}}Add Media{{/strong}} has moved to a new button.', {
```
I wonder if this wording is appropriate for users who signed up 3 days ago as well as those who signed up 3 years ago. The former might be confused by the "has moved" language. For them, it has always been that way. Maybe we should give this one to users who signed up before the change was made, and then give another tour ("btw, this is how you add media") to those who signed up after that date?

Copons replied:
Yup, definitely makes sense to only show this GT for users who signed up *before* the change was merged.
Do you know if there's already a function testing this, or should I do a manual check with `isUserOlderThan` the milliseconds since an arbitrary date I'll set right before merging?

lsinger commented:
Previous discussion now hidden because diff is outdated:

>> (@lsinger) I wonder if this wording is appropriate for users who signed up 3 days ago as well as those who signed up 3 years ago. The former might be confused by the "has moved" language. For them, it has always been that way. Maybe we should give this one to users who signed up before the change was made, and then give another tour ("btw, this is how you add media") to those who signed up after that date?

> (@Copons) Yup, definitely makes sense to only show this GT for users who signed up before the change was merged.
> Do you know if there's already a function testing this, or should I do a manual check with `isUserOlderThan` the milliseconds since an arbitrary date I'll set right before merging?

I've added a function `hasUserRegisteredBefore(date)` in #9974 that you can use once that is merged and you've rebased.

Copons commented:
@lsinger I've changed the close button text from "Close" to "Got it", following the same approach found in [another tour](https://github.com/Automattic/wp-calypso/blob/master/client/layout/guided-tours/tours/theme-sheet-welcome-tour.js#L111).

lsinger commented:
LGTM! 🚢 

Note: I update the tour version and date in the trigger to today.
